Aplib.Core/Desire/Goal.cs

```diff
@@ -1,59 +1,86 @@
+        public delegate Heuristics HeuristicFunction();
+
 
         /// <summary>
-        /// The <see cref="Tactic"/> used to achieve this <see cref="Goal"/>, which is executed during every iteration
-        /// of the BDI cycle.
-        /// </summary>
-        /// <seealso cref="Iterate()"/>
-        private readonly Tactic _tactic;
-        /// <summary>
-        /// The <see cref="IGoalPredicate"/> used to test whether this <see cref="Goal"/> has been completed.
+        /// Gets the <see cref="Heuristics"/> of the current state of the game.
         /// </summary>
-        /// <seealso cref="IsCompleted()"/>
-        private readonly IGoalPredicate _goalPredicate;
+        /// <remarks>If no heuristics have been calculated yet, they will be calculated first.</remarks>
+        public virtual Heuristics CurrentHeuristics => _currentHeuristics ??= heuristicFunction.Invoke();
 
-        // MetaData useful for debugging
         /// <summary>
         /// The name used to display the current goal during debugging, logging, or general overviews.
         /// </summary>
-        public readonly string Name;
+        public string Name { get; }
+
         /// <summary>
         /// The description used to describe the current goal during debugging, logging, or general overviews.
         /// </summary>
-        public readonly string Description;
+        public string Description { get; }
+
+
+        /// <summary>
+        /// The concrete implementation of this Goal's <see cref="HeuristicFunction"/>. Used to test whether this goal is
+        /// completed.
+        /// </summary>
+        /// <seealso cref="Evaluate"/>
+        protected HeuristicFunction heuristicFunction;
 
         /// <summary>
-        /// A goal effectively combines a predicate with a tactic, and aims to meet the predicate by applying the tactic.
-        /// Goals are combined in a <see cref="GoalStructure"/>, and are used to prepare tests or do the testing.
+        /// The backing field of <see cref="Heuristics"/>.
+        /// </summary>
+        private Heuristics? _currentHeuristics;
+
+        /// <summary>
+        /// The <see cref="Tactic"/> used to achieve this <see cref="Goal"/>, which is executed during every iteration
+        /// of the BDI cycle.
+        /// </summary>
+        /// <seealso cref="Iterate()"/>
+        private readonly Tactic _tactic;
+
+        /// <summary>
+        /// Creates a new goal with specified arguments. Upon creation, the <see cref="CurrentHeuristics"/> will be
+        /// set to <see cref="Heuristics.Default"/>.
         /// </summary>
         /// <param name="tactic">The tactic used to approach this goal.</param>
-        /// <param name="goalPredicate">The predicate which defines whether a goal is reached</param>
+        /// <param name="heuristicFunction">The heuristicFunction which defines whether a goal is reached</param>
```

Comment:
Let's not use parameter names in the XML documentation: 
```suggestion
        /// <param name="heuristicFunction">The heuristic function which defines whether a goal is reached</param>
```